src/lib/core/random.c

```diff
@@ -97,3 +103,82 @@ random_bytes(char *buf, size_t size)
 	while (generated < size)
 		buf[generated++] = rand();
 }
+
+uint64_t
+real_random(void)
+{
+	uint64_t result;
+	random_bytes((char *)&result, sizeof(result));
+	return result;
+}
+
+/**
+ * Helper function for the xoshiro256++ pseudo random generator:
+ * rotate left.
+ */
+static inline uint64_t
+rotl(uint64_t x, int k)
+{
+	return (x << k) | (x >> (64 - k));
+}
+
+/**
+ * xoshiro256++ pseudo random generator.
+ * http://prng.di.unimi.it/
+ * State is initialized with random_bytes().
+ *
+ * It is fast and doesn’t fail any known statistical test.
+ * About 2 times faster than conventional LCG rand() and
+ * mersenne-twister algorithms. Also both of them do fail
+ * some statistical tests.
+ * Here are some other reasons to choose xoshiro over
+ * mersenne-twister: https://arxiv.org/pdf/1910.06437.pdf
+ */
+uint64_t
+xoshiro_random(void)
+{
+	const uint64_t result = rotl(state[0] + state[3], 23) + state[0];
+	const uint64_t t = state[1] << 17;
+	state[2] ^= state[0];
+	state[3] ^= state[1];
+	state[1] ^= state[2];
+	state[0] ^= state[3];
+	state[2] ^= t;
+	state[3] = rotl(state[3], 45);
+	return result;
+}
+
+const char *
+xoshiro_state_str(void)
+{
+	return tt_sprintf("%llu %llu %llu %llu", (unsigned long long)state[0],
+						 (unsigned long long)state[1],
+						 (unsigned long long)state[2],
+						 (unsigned long long)state[3]);
+}
+
+int64_t
+real_random_in_range(int64_t min, int64_t max)
+{
+	assert(max >= min);
+	uint64_t range = (uint64_t)max - min;
+	uint64_t mask = UINT64_MAX >> bit_clz_u64(range | 1);
+	uint64_t result;
+	do {
+		result = real_random() & mask;
+	} while (result > range);
```

Comment:
This cycle doesn't seem to be trivial (at least from the first sight). For instance, in swim_scaled_rand()/gen_rand_value() calculation was obvious. Could you please supply it with a comment?

Reply:
Okay, to understand how it is supposed to work one should read this article https://www.pcg-random.org/posts/bounded-rands.html

Reply:
Yes, this  article is mentioned in my comment for the issue https://github.com/tarantool/tarantool/issues/5075#issuecomment-701052530
Should I also specify it in comment among the code?

Reply:
For sure, link to the article and brief explanation (in case link becomes invalid/outdated).

Reply:
Added